include/swift/AST/CASTBridging.h

```diff
@@ -112,6 +133,29 @@ typedef enum ENUM_EXTENSIBILITY_ATTR(open) BridgedMacroDefinitionKind : long {
   BridgedBuiltinExternalMacro
 } BridgedMacroDefinitionKind;
 
+typedef enum ENUM_EXTENSIBILITY_ATTR(open) ASTNodeKind : long {
+  ASTNodeKindExpr,
+  ASTNodeKindStmt,
+  ASTNodeKindDecl
+} ASTNodeKind;
+
+typedef struct BridgedASTNode {
+  void *ptr;
+  ASTNodeKind kind;
+} BridgedASTNode;
+
+typedef struct BridgedFuncDecl {
+  BridgedDeclContext declContext;
```

Comment:
`ASTGenVisitor` stores a raw decl context pointer, so `*_create` functions for decls that are also decl contexts split the resulting pointer into upcast `Decl` and `DeclContext` pointers so that the latter can be passed down to other `*_create` functions. We could pass around the raw decl pointer as the decl context but then we’d have to rely on these bridged constructors to first statically cast it to `Decl *` and then do a dynamic cast to `DeclContext *`, which sounds counterintuitive because `Decl` and `DeclContext` are unrelated classes.